### PR TITLE
Use upper/lowercase colors in Inline Color Editor

### PR DESCRIPTION
**Pull request by le717**
For #9519.
### Features
- Preserves case between value conversions and slider changes
- Related code updated to be case-insensitive
- New tests for conversion, with all existing tests passing
- Also made some minor code style fixes

This code is rather confusing (as already documented), so it was a bit of a challenge to get this written up. I once spent 30 digging around in various places for a one-line fix.

Now for the issues present here:
- This requires an update to [TinyColor v1.1.0+](https://github.com/bgrins/TinyColor/releases/tag/1.1.0), included with this PR. The creator refactored the library quite a bit since the v0.9.10 version currently used. TinyColor also unconditionally converts input to lowercase, which would make us have to keep track if the input is uppercase or not (which, in line with the previously documented issues, would be very hard because of the changing types). I have worked with Brian on the aforementioned patch to store the original input, which in turn makes this patch much easier to work out.

### CODE REVIEW

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 1]
```diff
@@ -217,10 +229,10 @@ define(function (require, exports, module) {
         var handler,
             _this = this;
```
Comment (MarcelGerber) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Just for consistency (I know it's not something you've changed), we usually call this variable `self`.

**Comment by redmunds** [comment]
@le717 This looks good, so you can work on the unit tests now.

**Comment by le717** [comment]
@redmunds Is there a test suite I can look at for reference? It seems that no matter what I try, I cannot get the test to set `uppercaseColors` to `true`.

**Comment by MarcelGerber** [comment]
The [FileFilters test suite](https://github.com/adobe/brackets/blob/master/test/spec/FileFilters-test.js) uses some preferences.

**Comment by redmunds** [comment]
@le717 Take a look in Editor-test.js at the "with soft tabs preference off" test.

**Comment by le717** [comment]
Neither of those are working. :\ Is there a dependency or state that needs to be loaded/enabled first?

File: src/extensions/default/InlineColorEditor/unittests.js [hunk 2]
```diff
@@ -473,34 +473,77 @@ define(function (require, exports, module) {
                     expect(colorEditor.getColor().toString()).toBe(result);
                 }
                 
-                it("should convert a hex color to rgb when mode button clicked", function () {
+                // Use uppercase colors
+                PreferencesManager.set("uppercaseColors", true);
```
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/unittests.js]: Code should only be executed inside `it()` callback functions to keep each test independent.

Here's what you should do:

1. Copy entire *original* `describe("conversions", function () {` block and give it a unique name such as `describe("conversions in UPPER CASE", function () {`

2. At top of `describe()` block and before first `it()` call, define these callbacks:

```js
                beforeEach(function () {
                    // Enable uppercase colors
                    PreferencesManager.set("uppercaseColors", true);
                });
                afterEach(function () {
                    // Re-disable uppercase colors
                    PreferencesManager.set("uppercaseColors", false);
                });
```

This will automatically set preference before each test and cleanup after each test in that `describe()` block.

Note: the `testConvert()` function will be defined in both the original and new `describe()` blocks, so you can move it up 1 level so there's only a single shared copy.

File: src/extensions/default/InlineColorEditor/unittests.js [hunk 3]
```diff
@@ -473,34 +473,77 @@ define(function (require, exports, module) {
...
+                PreferencesManager.set("uppercaseColors", false);
+                
+                it("should use lowercase colors", function () {
+                   expect(PreferencesManager.get("uppercaseColors")).toBe(false);
+                });
+                it("should convert an lowercase hex color to rgb when mode button clicked", function () {
                     testConvert("#112233", "rgba", "rgb(17, 34, 51)");
                 });
-                it("should convert a hex color to hsl when mode button clicked", function () {
+                it("should convert an lowercase hex color to hsl when mode button clicked", function () {
```
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/unittests.js]: It looks like all of these test are lowercase to lowercase or uppercase to uppercase -- I think there should be some tests for lowercase to uppercase and uppercase to lowercase, too.

Also, what is the expectation when original text is mixed case? I think there should be some tests added for that.

**Comment by le717** [comment]
@redmunds All changes except mixed-case tests made. However, the uppercase tests are still failing despite `uppercaseColors` confirmed to be `true`. I'm thinking it is an implementation error. I'll comment where I think it is coming in, but I am unsure of a fix ATM.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 1 above)
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: I'll change that to `self` next push.

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 4]
```diff
@@ -64,6 +67,7 @@ define(function (require, exports, module) {
         this._color = tinycolor(color);
         this._originalColor = color;
         this._redoColor = null;
+        this._isUpperCase = PreferencesManager.get("uppercaseColors");
```
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: I think the error is coming in here. Notice how `this._color` is also in the prototype, but this is not. I have gone back and forth including it there too or not, but I've always hit issues either way. I have a feeling that somehow, the unit tests editor is not picking up the updated `this._isUpperCase` value, and including it in the prototype as well will fix it. My question then is how to ensure listening to the `change` will properly update the prototype and not the object.

Yea, the inline color editor is rather a mess...

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 1 above)
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Fixed.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 4 above)
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @le717 The problem is in the test in this line of code in `testConvert()`:

```js
    expect(colorEditor.getColor().toString()).toBe(result);
```

`colorEditor.getColor()` returns a tinycolor with `_originalInput` correctly stored as UPPER case. But tinycolor `toString()` seems to ignore `_originalInput`, so I think this is a bug in tinycolor.

Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @le717 As a workaround, I think that line of `testConvert()` could be changed to:

```js
    expect(colorEditor.getColor().getOriginalInput()).toBe(result);
```

File: src/extensions/default/InlineColorEditor/unittests.js [hunk 5]
```diff
@@ -506,6 +506,53 @@ define(function (require, exports, module) {
...
+                });
+                afterEach(function () {
+                    // Re-disable uppercase colors
+                    PreferencesManager.set("uppercaseColors", false);
+                });
+                
+                it("should use uppercase colors", function () {
+                    expect(PreferencesManager.get("uppercaseColors")).toBe(true);
+                });
+                it("should convert a hex color to rgb when mode button clicked", function () {
```
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/unittests.js]: Each test should have a unique name, so you'll need to add something like "in uppercase" to each of these duplicated tests.

**Comment by le717** [comment]
@redmunds Test names, mixed case tests, and use `getOriginalInput()` changes pushed, all tests now pass. I didn't get your comment saying to use `getOriginalInput()` until later, so when I read your first one my thought was to use that method.

If everything looks good, I'll flatten this.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 4 above)
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: `tinycolor.toString()` actually works as the author intended. It does not ignore the original case, but rather normalizes (to lowercase) it before returning it. There was actually no way to get the original input (it wasn't even stored) until I suggested it and sent a patch in the form of `tinycolor.getOriginalInput()`.

Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Yes, but since you added `tinycolor.getOriginalInput()`, I think it changes expectations. I would expect that method to be called something like `toNormalizedString()`. I'll leave it up to you if you want to pursue that.

**Comment by redmunds** [comment]
@le717 Looks good! Squash it.

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 6]
```diff
@@ -64,6 +67,10 @@ define(function (require, exports, module) {
...
+        PreferencesManager.on("change", "uppercaseColors", function () {
+            this._isUpperCase = PreferencesManager.get("uppercaseColors");
+        }.bind(this));
```
Comment (peterflynn) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Drive-by comment: this leaks listeners.  Each time a color editor is opened, another change listener is added to PreferencesManager.  The change listener will also keep the entire color editor in memory forever, since it's part of its closure.  We shouldn't merge until this is fixed.

Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Good catch! It sucks that JS objects have a constructor but not a destructor.

@le717 You could add a `ColorEditor.destroy()` method to remove the listener, which could be called from the `InlineColorEditor.onClosed()` method.

Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @redmunds Doing so always throws "undefined is not a function", using both `ColorEditor.destroy()` and `this.colorEditor.destroy()` (like surrounding code).

Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Make sure new method is part of object prototype: `ColorEditor.prototype.destroy()`

Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Sorry, I misread that (my thoughts are a bit scrambled right now with classes just starting back). Fixed.

**Comment by le717** [comment]
@peterflynn @redmunds Listener hopefully patched, hopefully passes your inspection. Already squashed the commits.

File: src/extensions/default/InlineColorEditor/unittests.js [hunk 7]
```diff
@@ -503,6 +503,65 @@ define(function (require, exports, module) {
                 it("should convert an hsla color to rgba when mode button clicked", function () {
                     testConvert("hsla(152, 12%, 22%, 0.7)", "rgba", "rgba(49, 63, 57, 0.7)");
                 });
+                it("should convert a mixed case hsla color to rgba when mode button clicked", function () {
```
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/unittests.js]: Thanks for the mixed case tests!

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 8]
```diff
@@ -215,12 +234,12 @@ define(function (require, exports, module) {
...
         var handler,
-            _this = this;
+            self = this;
         handler = function (event) {
-            var colorObject, newColor, newFormat;
-            newFormat = $(event.currentTarget).html().toLowerCase().replace("%", "p");
-            newColor = _this.getColor();
-            colorObject = tinycolor(newColor);
+            var newFormat   = $(event.currentTarget).html().toLowerCase().replace("%", "p"),
+                newColor    = self.getColor().toString(),
```
Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @le717 All of the test *currently* pass, but I'm wondering if all calls to  `TinyColor.toString()` should be removed from the unit tests since we know it's not case-sensitive?

**Comment by redmunds** [comment]
Done with review. This looks good. Just [one last question](https://github.com/adobe/brackets/pull/9596#discussion_r22954092) about unit tests.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 8 above)
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @redmunds That's a good question. As shown in this line you commented on, there are still uses of `toString()` in the extension itself as well as in the unit tests. I have run into some confusion in this PR over which to use, just to note that. As for unit tests, all I can do is test it using `getOriginalInput()`, but I'm not that it will fully work as the latter is not a drop-in replacement for `toString()`.

Comment (redmunds) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @le717 Oops -- I only meant that it should be changed in unit tests.

Ignoring case, I think the only time that `toString()` returns something different from `getOriginalInput()` is between the time an change is made in inline color editor and `ColorEditor._commitColor()` is called, right? If so, then `getOriginalInput()` can always be used in unit tests.

But, that's not true in ColorEditor since we need to know when to convert the case returned from `toString()`.

I'm ok with that, but another solution would be to add a new method something like `ColorEditor.prototype.getColorString()` that calls `TinyColor.toString()` and then converts the case, if necessary.

Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: > Ignoring case, I think the only time that ` toString()` returns something different from `getOriginalInput()` is between the time an change is made in inline color editor and `ColorEditor._commitColor()` is called, right? If so, then `getOriginalInput()` can always be used in unit tests.

That is correct. 

> I'm ok with that, but another solution would be to add a new method something like `ColorEditor.prototype.getColorString()` that calls `TinyColor.toString()` and then converts the case, if necessary.

If you recall, in the toggle-button implementation I actually had method by that name which did just that. However, I removed it in the refactoring. I can always restore it, but that may be better suited for a different PR.

**Comment by le717** [comment]
Questions answered, switched unit tests to use `getOriginalInput()`. Ready on your command.

**Comment by redmunds** [comment]
Looks good. Merging.

@le717 Don't forget to add new preference to https://github.com/adobe/brackets/wiki/How-to-Use-Brackets#list-of-supported-preferences Note that prefs are now in alphabetic order.

**Comment by le717** [comment]
@redmunds Done. I wanted to add that this has the chance of breaking extension _if_ they for whatever reason use the TinyColor version from here (which I don't think they do, they host their own) because of the API changes, so it might be good to list it as a library update in the change log.

**Comment by MarcelGerber** [comment]
I just searched through all the extensions on the registry and luckily, there's not a single one using tinycolor. So we're pretty much good to go.
